pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RangeIndexRule.java

```diff
@@ -0,0 +1,122 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Create range-index for columns used in <, > and between predicate
+ * Skip the inverted and sorted index columns if recommended
+ */
+public class RangeIndexRule extends AbstractRule {
+  private final Logger LOGGER = LoggerFactory.getLogger(RangeIndexRule.class);
+  private final RangeIndexRuleParams _params;
+
+  public RangeIndexRule(InputManager input, ConfigManager output) {
+    super(input, output);
+    _params = input.getRangeIndexRuleParams();
+  }
+
+  @Override
+  public void run() {
+    int numCols = _input.getNumCols();
+    double[] weights = new double[numCols];
+    AtomicDouble totalWeight = new AtomicDouble(0);
+
+    // For each query, find out the range columns
+    // and accumulate the (weighted) frequencies
+    _input.getParsedQueries().forEach(query -> {
+      Double weight = _input.getQueryWeight(query);
+      totalWeight.addAndGet(weight);
+      FixedLenBitset fixedLenBitset = parseQuery(_input.getQueryContext(query));
+      LOGGER.debug("fixedLenBitset {}", fixedLenBitset);
+      for (Integer i : fixedLenBitset.getOffsets()) {
+        weights[i] += weight;
+      }
+    });
+    LOGGER.debug("Weight: {}, Total {}", weights, totalWeight);
+
+    for (int i = 0; i < numCols; i++) {
+      String colName = _input.intToColName(i);
+      // This checks if column is not already recommended for inverted index or sorted index
+      // As currently, only numeric columns are selected in range index creation, we will skip non numeric columns
+      if (((weights[i] / totalWeight.get()) > _params.THRESHOLD_MIN_PERCENT_RANGE_INDEX) &&
+          !_output.getIndexConfig().getSortedColumn().equals(colName) &&
+          !_output.getIndexConfig().getInvertedIndexColumns().contains(colName) &&
+          _input.getFieldType(colName).isNumeric()) {
+        _output.getIndexConfig().getRangeIndexColumns().add(colName);
+      }
+    }
+  }
+
+  public FixedLenBitset parseQuery(QueryContext queryContext) {
+    if (queryContext.getFilter() == null) {
+      return FixedLenBitset.IMMUTABLE_EMPTY_SET;
+    }
+
+    LOGGER.trace("Parsing Where Clause: {}", queryContext.getFilter().toString());
+    return parsePredicateList(queryContext.getFilter());
+  }
+
+  /**
+   * As cardinality does not matter, AND and OR columns will be considered
+   * @param filterContext filterContext
+   * @return FixedLenBitset for range predicates in this query
+   */
+  private FixedLenBitset parsePredicateList(FilterContext filterContext) {
+    FilterContext.Type type = filterContext.getType();
+    FixedLenBitset ret = MUTABLE_EMPTY_SET();
+    if (type == FilterContext.Type.AND) {
+      for (int i = 0; i < filterContext.getChildren().size(); i++) {
+        FixedLenBitset childResult = parsePredicateList(filterContext.getChildren().get(i));
+        ret.union(childResult);
+      }
+    } else if (type == FilterContext.Type.OR) {
```

Comment:
Look like we can merge the two if branch of L97 and L102 as one:
```
if (type == FilterContext.Type.AND || type == FilterContext.Type.OR)
```

Reply:
updated!